quantum/process_keycode/process_autocorrect.c

```diff
@@ -0,0 +1,201 @@
+                typo_buffer_size = 0;
+                return false;
+            }
+    }
+    return true;
+}
+
+/**
+ * @brief Process handler for autocorrect feature
+ *
+ * @param keycode Keycode registered by matrix press, per keymap
+ * @param record keyrecord_t structure
+ * @return true Continue processing keycodes, and send to host
+ * @return false Stop processing keycodes, and don't send to host
+ */
+
+bool process_autocorrect(uint16_t keycode, keyrecord_t *record) {
+    static uint8_t typo_buffer[AUTOCORRECT_MAX_LENGTH] = {KC_SPC};
+    static uint8_t typo_buffer_size                    = 1;
+#ifndef NO_ACTION_ONESHOT
+    const uint8_t mods = get_mods() | get_oneshot_mods();
+#else
+    const uint8_t mods = get_mods();
+#endif // NO_ACTION_ONESHOT
+
+    if (keycode == AUTOCRT) {
+        if (record->event.pressed) {
+            typo_buffer_size = 0;
+            keymap_config.autocorrect_enable ^= 1;
+            eeconfig_update_keymap(keymap_config.raw);
+        }
+        return false;
+    }
+
+    if (!keymap_config.autocorrect_enable) {
+        typo_buffer_size = 0;
+        return true;
+    }
+
+    if (!record->event.pressed) {
+        return true;
+    }
+
+    if (!process_autocorrect_user(&keycode, record, &typo_buffer_size, mods)) {
+        return true;
+    }
+    // Subtract buffer for Backspace key, reset for other non-alpha.
+    if (keycode == KC_QUOT) {
+        // Treat " (shifted ') as a word boundary.
+        if ((mods & MOD_MASK_SHIFT) != 0) {
+            keycode = KC_SPC;
+        }
+    } else if (!(KC_A <= keycode && keycode <= KC_Z)) {
+        if (keycode == KC_BSPC) {
+            // Remove last character from the buffer.
+            if (typo_buffer_size > 0) {
+                --typo_buffer_size;
+            }
+            return true;
+        } else if (KC_1 <= keycode && keycode <= KC_SLSH && keycode != KC_ESC) {
+            // Set a word boundary if space, period, digit, etc. is pressed.
+            // Behave more conservatively for the enter key. Reset, so that enter
+            // can't be used on a word ending.
+            if (keycode == KC_ENT || (keycode == KC_MINUS && (get_mods() | get_oneshot_mods()) & MOD_MASK_SHIFT)) {
+                typo_buffer_size = 0;
+            }
+            keycode = KC_SPC;
+        } else {
+            // Clear state if some other non-alpha key is pressed.
+            typo_buffer_size = 0;
+            return true;
+        }
+    }
+
+    // Rotate oldest character if buffer is full.
+    if (typo_buffer_size >= AUTOCORRECT_MAX_LENGTH) {
+        memmove(typo_buffer, typo_buffer + 1, AUTOCORRECT_MAX_LENGTH - 1);
+        typo_buffer_size = AUTOCORRECT_MAX_LENGTH - 1;
+    }
+
+    // Append `keycode` to buffer.
+    typo_buffer[typo_buffer_size++] = keycode;
+    // Return if buffer is smaller than the shortest word.
+    if (typo_buffer_size < AUTOCORRECT_MIN_LENGTH) {
+        return true;
+    }
```

Comment:
@getreuer Your original code zeroes the next element if the buffer is not full. Is it necessary to avoid spurious matches?

https://github.com/getreuer/qmk-keymap/blob/main/features/autocorrection.c#L126-L127

Reply:
Thanks for pointing that out! I agree with removing that zero as drashna did. I should clean this up in my repo. That zero shouldn't have an effect since the for loop accesses the buffer only at i = typo_buffer_size - 1 down to 0.

While I was developing autocorrection, I considered a few different approaches to implementing the trie lookup. I don't remember what specifically motivated the trailing zero, but it's probably left over cruft from an earlier implementation. Either that or a C habit of null-terminated strings =)